File: frontend/src/types/simulation.ts
export interface ClimateScenario {
  id: string;
  code: string;
  name: string;
  pathway: string;
  description: string;
  temp_anomaly_c: number;
  precip_factor: number;
  co2_ppm: number;
}

export interface PlantSpecies {
  id: string;
  name: string;
  scientific_name: string;
  crop_type: string;
  base_kc: number;
  max_root_depth_cm: number;
  optimal_temp_c: number;
}

export interface HRU {
  id: string;
  hru_number: number;
  land_use: string;
  soil_type: string;
  curve_number_ii: number;
  area_fraction: number;
  plant_species?: PlantSpecies | null;
}

export interface Subbasin {
  id: string;
  subbasin_number: number;
  name: string;
  area_km2: number;
  mean_slope_percent: number;
  hrus: HRU[];
}

export interface Watershed {
  id: string;
  code: string;
  name: string;
  country: string;
  area_km2: number;
  elevation_min_m: number;
  elevation_max_m: number;
  outlet_lat: number;
  outlet_lon: number;
  dem_metadata?: any;
  subbasins: Subbasin[];
}

export interface SimulationResult {
  day_index: number;
  date_str: string;
  precip_mm: number;
  temp_c: number;
  solar_rad_mj: number;
  potential_et_mm: number;
  actual_et_mm: number;
  surface_runoff_mm: number;
  percolation_mm: number;
  streamflow_m3s: number;
  soil_moisture_vol: number;
  plant_transpiration_mm: number;
  root_water_uptake_mm: number;
  cwsi_stress_index: number;
  sap_flow_velocity_cmh: number;
}

export interface SimulationRun {
  id: string;
  user_id: string;
  watershed_id: string;
  scenario_id: string;
  name: string;
  status: "PENDING" | "RUNNING" | "COMPLETED" | "FAILED";
  duration_days: number;
  irrigation_efficiency: number;
  scenario?: ClimateScenario;
  summary_metrics?: {
    total_precip_mm: number;
    total_surface_runoff_mm: number;
    total_actual_et_mm: number;
    total_discharge_hm3: number;
    peak_streamflow_m3s: number;
    mean_cwsi: number;
    drought_stress_status: string;
  };
  created_at: string;
}

export interface TwinWebSocketTick {
  type: "TWIN_STATE_TICK" | "ERROR";
  day_index: number;
  date: string;
  weather: {
    precip_mm: number;
    temp_c: number;
    solar_rad_mj: number;
  };
  micro_plant: {
    transpiration_mm: number;
    root_water_uptake_mm: number;
    cwsi_stress_index: number;
    sap_flow_velocity_cmh: number;
  };
  meso_soil: {
    soil_moisture_vol: number;
    soil_water_depth_mm: number;
    percolation_mm: number;
  };
  macro_watershed: {
    surface_runoff_mm: number;
    streamflow_m3s: number;
    actual_et_mm: number;
  };
}

File: backend/app/schemas/simulation.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, ConfigDict

# --- Climate Scenarios ---
class ClimateScenarioResponse(BaseModel):
    id: str
    code: str
    name: str
    pathway: str
    description: str
    temp_anomaly_c: float
    precip_factor: float
    co2_ppm: float
    model_config = ConfigDict(from_attributes=True)

# --- Plant Species ---
class PlantSpeciesResponse(BaseModel):
    id: str
    name: str
    scientific_name: str
    crop_type: str
    base_kc: float
    max_root_depth_cm: float
    optimal_temp_c: float
    model_config = ConfigDict(from_attributes=True)

# --- Watershed & HRUs ---
class HRUResponse(BaseModel):
    id: str
    hru_number: int
    land_use: str
    soil_type: str
    curve_number_ii: float
    area_fraction: float
    plant_species: Optional[PlantSpeciesResponse] = None
    model_config = ConfigDict(from_attributes=True)

class SubbasinResponse(BaseModel):
    id: str
    subbasin_number: int
    name: str
    area_km2: float
    mean_slope_percent: float
    hrus: List[HRUResponse] = []
    model_config = ConfigDict(from_attributes=True)

class WatershedResponse(BaseModel):
    id: str
    code: str
    name: str
    country: str
    area_km2: float
    elevation_min_m: float
    elevation_max_m: float
    outlet_lat: float
    outlet_lon: float
    dem_metadata: Optional[Dict[str, Any]] = None
    subbasins: List[SubbasinResponse] = []
    model_config = ConfigDict(from_attributes=True)

# --- Simulation Run ---
class SimulationRunCreate(BaseModel):
    name: str
    watershed_id: str
    scenario_id: str
    duration_days: int = 365
    irrigation_efficiency: float = 0.85
    parameters: Optional[Dict[str, Any]] = None

class SimulationResultResponse(BaseModel):
    day_index: int
    date_str: str
    precip_mm: float
    temp_c: float
    solar_rad_mj: float
    potential_et_mm: float
    actual_et_mm: float
    surface_runoff_mm: float
    percolation_mm: float
    streamflow_m3s: float
    soil_moisture_vol: float
    plant_transpiration_mm: float
    root_water_uptake_mm: float
    cwsi_stress_index: float
    sap_flow_velocity_cmh: float
    model_config = ConfigDict(from_attributes=True)

class SimulationRunResponse(BaseModel):
    id: str
    user_id: str
    watershed_id: str
    scenario_id: str
    name: str
    status: str
    duration_days: int
    irrigation_efficiency: float
    scenario: Optional[ClimateScenarioResponse] = None
    summary_metrics: Optional[Dict[str, Any]] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)
